ONSdigital/ssdc-rm-ddl
# 452 Add rollback script and tests

# Has the DDL schema changed?
*Pull requests which include schema changes should be labelled with the `schema change` label for visibility*

*The [common entity model pom.xml](ssdc-rm-common-entity-model/pom.xml) file `project.version` must be bumped
appropriately if there are any common entity code changes*

~~* [ ] This PR includes DDL schema changes and has been labelled correctly, bumping to new schema
  version:~~ 
~~* [ ] The POM has been updated with an appropriate version bump if required~~

# Motivation and Context

We need a mechanism for rolling back a patch migration if needed.

This PR adds a python database rollback tool which looks up the applied patches record and can run the associated rollback script for the last `n` patches, while also updating the patches and version record appropriately.

# What has changed



- Add rollback script
- Add tests
- Update README

# How to test?



Check out this branch and add a dummy migration patch and bump the `CURRENT_VERSION` in the patch database script locally. If you do not add a matching rollback patch, the `make test-patches` run should fail. If you write a rollback (which just has to be a valid SQL file, as warned in the README, the test does not verify the contents of the patches/rollbacks at all) then the tests should run the rollback and pass.

# Links



https://trello.com/c/McV3gYiI/452-spike-ddl-rollback-process-1-day

Travis build with the test failing because of missing rollback (dummy patches): https://app.travis-ci.com/github/ONSdigital/ssdc-rm-ddl/jobs/586718407#L390

Travis build passing running dummy patches: https://app.travis-ci.com/github/ONSdigital/ssdc-rm-ddl/jobs/586720299#L347
